### PR TITLE
Add framework agnostic MDX types

### PR DESCRIPTION
**Pull request by remcohaszing**
This is compatible with `@mdx-js/mdx` v2 and `xdm`.

The types are based on the assumption a global JSX namespace is used. This holds true for React, Preact, and any framework that uses jsx import source.

The tests will fail, because `dtslint` won’t let me configure `jsxImportSource`, causing the tests to pull in `react`, which conflicts with the JSX test implementation.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/mdx/tsconfig.json
Comment (ChristianMurphy) [inline types/mdx/tsconfig.json:22]: Interestingly enough this seems to error

```
Error: mdx: Only index.d.ts may be listed explicitly in tsconfig's "files" entry.
Other d.ts files must either be referenced through index.d.ts, tests, or added to OTHER_FILES.txt.
```

**Review by ChristianMurphy — COMMENTED** [review]
Thanks Remco!

General question.
What is the thinking having this in DT vs having it in the mdx repo directly?

**Review by wooorm — COMMENTED** [review]
- Per Christian’s question, I also think we could do it and `@mdx-js/types` or so project
- `props.components` can perhaps be typed:
  - Normal HTML (`<div>`):
    ```ts
    ComponentType<Props> = // To do: JSX.ElementClass || () => JSX.Element

    Partial<{
      [TagName in keyof JSX.IntrinsicElements]:
        | keyof JSX.IntrinsicElements
        | ComponentType<JSX.IntrinsicElements[TagName]>
    }>
    ```
  - custom components (`<YouTube>`), something along the lines of:
    ```ts
    Record<string, keyof JSX.IntrinsicElements | ComponentType<any>>
    ```
  - objects (`<object.field>`)
- Other than an `export default`, anything else can be exported:
  ```ts
  {[exports: string]: unknown}
  ```

**Comment by remcohaszing** [comment]
> The tests will fail, because `dtslint` won’t let me configure `jsxImportSource`, causing the tests to pull in `react`, which conflicts with the JSX test implementation.

I figured I can work around this by not using `jsxImportSource`, but a stub `jsxFactory`. As far as TypeScript is concerned, those are equivalent as far as a global JSX namespace type definition goes.

> What is the thinking having this in DT vs having it in the mdx repo directly?

These are types for MDX **usage**. The various MDX tools contain **build** tools and types for them.

When using MDX, the build tool types might not even be included. They probably shouldn’t even be. The MDX is likely to be used within a web context, whereas the build tools probably use NodeJS APIs. When using an MDX build tool, this doesn’t mean `*.mdx` can be imported from within the build tool. Also the other way around, if a build tool is imported for the sole purpose of making `*.mdx` imports available, this could have side effects. I.e. the build tool is likely to depend on `@types/node`, which is preverably avoided inside a web only project.

I don’t have strong feelings on where the types should live, just that they should be in a separate package. I think it makes sense to define this in `@types/mdx`. This is fairly similar to `css-loader` and `@types/css-modules`. It also makes sense to me to publish them as `@mdx-js/types` as @wooorm proposed.

---

I think we should decide where to track the types before I address the other issues.

File: types/mdx/tsconfig.json
Comment (orta) [reply types/mdx/tsconfig.json:22]: `md.d.ts` won't get picked up by the TypeScript file resolver unless it's referenced somewhere, other files means someone might import it via `module/filename` in their code

**Comment by orta** [comment]
I think my gut says that here in DT probably a better call. With a bit of import trickery that could also mean JS users get the changes without having to add the dependency via ATA

**Comment by wooorm** [comment]
@orta forgive me, but what is ATA? And what would that trickery be to get that working?

@ChristianMurphy any reservations?

**Comment by ChristianMurphy** [comment]
I'm fine with it being on DT, just wanted to check if that was intended/best-practice.

File: types/mdx/tsconfig.json
Comment (remcohaszing) [reply types/mdx/tsconfig.json]: I imported it in the tests instead.

**Comment by remcohaszing** [comment]
I updated the type tests to work with the classic JSX transform, as it doesn’t make a difference for the type definitions.

The `*.mdx` module now defines a default export only. I also added more detailed instructions on how module augmentation works. The difference between TypeScript modules and scripts has been confusing to me and several people I know, so I felt it needed explanation if this is how people can augment MDX types.

I also tried the following, but it didn’t work:

```ts
declare module '*.mdx' {
  type MDXModule = Record<string, unknown> & {
      default: MDXComponent;
  };

  const mdx: MDXModule;
  type MDXComponent = (props: Record<string, unknown>) => JSX.Element;
  export = mdx;
}
```

File: types/mdx/tslint.json
Comment (ChristianMurphy) [inline types/mdx/tslint.json:6]: would it be possible to inline these in the places it is needed?

> The linter configuration file, tslint.json should contain { "extends": "dtslint/dt.json" }, and no additional rules.
>
> If for some reason some rule needs to be disabled, disable it for that specific line using // tslint:disable-next-line:[ruleName] — not for the whole package, so that disabling can be reviewed. (There are some legacy lint configs that have additional contents, but these should not happen in new work.)

https://github.com/DefinitelyTyped/DefinitelyTyped#linter-tslintjson

**Comment by peterblazejewicz** [comment]
> , but what is ATA? And what would that trickery be to get that working?

@wooorm
Automatic type acquisition (or automatic type resolution):
https://code.visualstudio.com/docs/nodejs/working-with-javascript#_typings-and-automatic-type-acquisition
It applies to all IDEs IMO (or ones supporting js/ts)

File: types/mdx/index.d.ts
Comment (wooorm) [inline types/mdx/index.d.ts:59]: Should we make this smarter?

I tried to type it like this yesterday: https://github.com/wooorm/xdm/blob/db77de625e018257631543591d5794a2e88952bb/complex-types.d.ts#L54. The benefit is that `props.components` is typed.
The downside is that not every JSX runtime supports components / JSX.Element properly — Preact had some problems (React / Vue were fine).

**Comment by wooorm** [comment]
> ```ts
> declare module '*.mdx' {
>   type MDXModule = Record<string, unknown> & {
>       default: MDXComponent;
>   };
> 
>   const mdx: MDXModule;
>   type MDXComponent = (props: Record<string, unknown>) => JSX.Element;
>   export = mdx;
> }
> ```
> — @remcohaszing

Yeah, unfortunately It indeed doesn’t work. I)t’s also more broken I believe. `export = x` sets `export default` to `x`. So it creates an extra `default`

File: types/mdx/index.d.ts
Comment (remcohaszing) [reply types/mdx/index.d.ts:63]: Does this suffice?

```ts
declare module '*.mdx' {
    type MDXComponentsProp = Record<
        string,
        JSX.ElementClass | keyof JSX.IntrinsicElements | ((props: Record<string, unknown>) => JSX.Element)
    >;
    type MDXProps = Record<string, unknown> & { components?: MDXComponentsProp };

    export default function MDXContent(props: MDXProps): JSX.Element;
}
```

Although this would be nice, `JSX.ElementClass` and `JSX.IntrinsicElements` don’t have to be supported by a JSX runtime. Can we detect if a type exists on a runtime?

I tried the following, but it says JSX can’t be used as a type:

```ts
JSX extends { ElementClass: infer T } ? T : never
```

I also tried the following, but it says JSX can’t be used as a value:

```ts
typeof JSX extends { ElementClass: infer T } ? T : never
```

Ultimately I believe not specifying `components` (meaning they’re `unknown`) is better than specifying them incorrectly.

---

> The downside is that not every JSX runtime supports components / JSX.Element properly

Could you explain what exactly isn’t supported? The `JSX.Element` type is key to using JSX at all. Frameworks can support `createElement.JSX.Element` as well (where `createElement` represents the JSX pragma function), but JSX supports the automatic runtime, which relies on the global `JSX` namespace.

File: types/mdx/md.d.ts
Comment (remcohaszing) [reply types/mdx/md.d.ts:9]: I wanted to make this opt-in, because I imagine some people may want to use markdown files differently. I.e. use `.mdx` files using `@mdx-js/loader` and `.md` files using `file-loader`. This is highly subjective though. If you have strong feelings for merging them into `index.d.ts` that’s fine with me. The same goes for other markdown extensions.

Another option would be to not support `.md` at all here.

--- 

> This code seems to mean that, when `.mdx` files are augmented with more exports, those exports also apply to `.md` files. Is that correct, intentional, and should it be that way?

This is intentional, but also subjective.

File: types/mdx/index.d.ts
Comment (remcohaszing) [reply types/mdx/index.d.ts:54]: I think this should be self explanatory to TypeScript users. It works the same as for example `@types/node` and `@types/css-modules`.

File: types/mdx/tslint.json
Comment (remcohaszing) [reply types/mdx/tslint.json:6]: The rule `npm-naming` is reported on the first line of `index.d.ts`. If this is disabled, it precedes the header, meaning the header can’t be parsed.

Since `npm-naming` is already disabled and `no-single-declare-module` is reported twice, I think it’s ok to disable that rule here too. Also if we decide to merge the `.md` module declaration into `index.d.ts`, this rule will no longer be supported, so let’s resolve that discussion first. :)

**Review by ChristianMurphy — APPROVED** [review]
Thanks @remcohaszing !

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
@ChristianMurphy Can you give an explicit approval again, if you’re a-okay with this?

I think it’s now okay from the unified maintainers perspective and can receive a review from @peterblazejewicz and @orta.

**Review by ChristianMurphy — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]
LGMT!

**Pull request merged**
